### PR TITLE
feat: reco training script with multi paths sets

### PR DESCRIPTION
**Pull request by charlesmindee**
This PR adds the option to train recognition with multiple datasets.
the arg `data_path` in the script is replaced by a `train_data_path`, a list of 1 or more paths and `val_data_path`, a path.
The structure inside each path must be: `path/images/` and `path/labels.json`.
Any feedback is welcome!

### CODE REVIEW

File: references/recognition/train_pytorch.py [hunk 1]
```diff
@@ -210,7 +229,8 @@ def parse_args():
     parser = argparse.ArgumentParser(description='DocTR train text-recognition model (PyTorch)',
                                      formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 
-    parser.add_argument('data_path', type=str, help='path to data folder')
+    parser.add_argument('train_data_path', type=List[str], help='list of path(s) to train data folder')
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:232]: Are you positive that we're able to pass list of string to a bash script?

Comment (fg-mindee) [reply references/recognition/train_pytorch.py:232]: Perhaps keeping the same old path, and simply assuming that all parts of the training data are inside that folder?

File: references/recognition/train_pytorch.py [hunk 2]
```diff
@@ -131,6 +134,22 @@ def main(args):
         ]),
     )
 
+    # If multiple paths provided, merge datasets
+    if len(args.train_data_path) > 1:
+        for i in range(1, len(args.train_data_path)):
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:139]: for data_path in args.train_data_path:

File: references/recognition/train_pytorch.py [hunk 3]
```diff
@@ -131,6 +134,22 @@ def main(args):
...
+            train_set.merge_dataset(
+                RecognitionDataset(
+                    img_folder=os.path.join(args.train_data_path[i], 'images'),
+                    labels_path=os.path.join(args.train_data_path[i], 'labels.json'),
+                    sample_transforms=Compose([
+                        T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
+                        # Augmentations
+                        T.RandomApply(T.ColorInversion(), .1),
+                        ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.02),
+                    ]),
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:149]: with the merge strategy, no need to specify the transform of the next dataset parts

File: references/recognition/train_tensorflow.py [hunk 4]
```diff
@@ -135,6 +138,26 @@ def main(args):
...
+                    labels_path=os.path.join(args.train_data_path[i], 'labels.json'),
+                    sample_transforms=T.Compose([
+                        T.RandomApply(T.ColorInversion(), .1),
+                        T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
+                        # Augmentations
+                        T.RandomJpegQuality(60),
+                        T.RandomSaturation(.3),
+                        T.RandomContrast(.3),
+                        T.RandomBrightness(.3),
+                    ]),
```
Comment (fg-mindee) [inline references/recognition/train_tensorflow.py:157]: same here

File: references/recognition/train_tensorflow.py [hunk 5]
```diff
@@ -135,6 +138,26 @@ def main(args):
             T.RandomBrightness(.3),
         ]),
     )
+
+    # If multiple paths provided, merge datasets
+    if len(args.train_data_path) > 1:
+        for i in range(1, len(args.train_data_path)):
```
Comment (fg-mindee) [inline references/recognition/train_tensorflow.py:144]: same here

**Review by fg-mindee — COMMENTED** [review]
Thanks for the PR! I'm not sure passing a list to a bash script is a good idea. Say we have dozens of parts, it would be better to assume that all parts are the subfolders of `train_set` for instance (`train_set/part1`, `train_set/part2`, etc.). If the json labels are directly in this folder, then there is only 1 part, if not, we instantiate the set with the first part, then merge the others. What do you think?

File: references/recognition/train_pytorch.py [hunk 6]
```diff
@@ -119,17 +122,39 @@ def main(args):
...
+                ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.02),
+            ]),
+        )
+        for folder in os.listdir(args.train_data_path)[1:]:
+            train_set.merge_dataset(
+                RecognitionDataset(
+                    img_folder=os.path.join(args.train_data_path, folder, 'images'),
+                    labels_path=os.path.join(args.train_data_path, folder, 'labels.json'),
+                )
+            )
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:157]: Let's make this even shorter :
```python
base_path = Path(args.train_path)
parts = [base_path] if base_path.joinpath('labels.json').is_file() else [base_path.joinpath(sub) for sub in os.listdir(base_path)]
train_set = RecognitionDataset(
    parts[0].joinpath('images'),
    parts[0].joinpath('labels.json'),
    sample_transforms=Compose([
                T.Resize((args.input_size, 4 * args.input_size), preserve_aspect_ratio=True),
                # Augmentations
                T.RandomApply(T.ColorInversion(), .1),
                ColorJitter(brightness=0.3, contrast=0.3, saturation=0.3, hue=0.02),
        ]),
)
if len(parts) > 1:
    for subfolder in parts[1:]:
        train_set.merge_dataset(RecognitionDataset(subfolder.joinpath('images'), subfolder.joinpath('labels.json'))

```

File: references/recognition/train_pytorch.py [hunk 7]
```diff
@@ -210,7 +235,8 @@ def parse_args():
     parser = argparse.ArgumentParser(description='DocTR train text-recognition model (PyTorch)',
                                      formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 
-    parser.add_argument('data_path', type=str, help='path to data folder')
+    parser.add_argument('train_data_path', type=str, help='path to train data folder(s)')
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:238]: can you rename it to `train_path` please?

File: references/recognition/train_pytorch.py [hunk 8]
```diff
@@ -210,7 +235,8 @@ def parse_args():
...
+    parser.add_argument('val_data_path', type=str, help='path to val data folder')
```
Comment (fg-mindee) [inline references/recognition/train_pytorch.py:239]: val_path perhaps?

File: references/recognition/train_tensorflow.py [hunk 9]
```diff
@@ -230,6 +259,8 @@ def parse_args():
     parser = argparse.ArgumentParser(description='DocTR train text-recognition model (TensorFlow)',
                                      formatter_class=argparse.ArgumentDefaultsHelpFormatter)
 
+    parser.add_argument('train_data_path', type=str, help='path to train data folder(s)')
+    parser.add_argument('val_data_path', type=str, help='path to val data folder')
```
Comment (fg-mindee) [inline references/recognition/train_tensorflow.py:263]: same here

File: references/recognition/train_tensorflow.py [hunk 10]
```diff
@@ -121,20 +123,47 @@ def main(args):
...
+            ]),
+        )
+        for folder in os.listdir(args.train_data_path)[1:]:
+            train_set.merge_dataset(
+                RecognitionDataset(
+                    img_folder=os.path.join(args.train_data_path, folder, 'images'),
+                    labels_path=os.path.join(args.train_data_path, folder, 'labels.json'),
+                )
+            )
+
```
Comment (fg-mindee) [inline references/recognition/train_tensorflow.py:166]: same here

**Review by fg-mindee — COMMENTED** [review]
One last improvement suggestion and we're good to merge!

**Review by fg-mindee — DISMISSED** [review]
Thanks again!

**Review by fg-mindee — APPROVED** [review]
Thanks!
